### PR TITLE
Add hpa metrics collector

### PR DESCRIPTION
**Pull request by EmanekaT**
Sorry that's a lot of code. Go is redundant. 
Tested on kubestage: https://fluffy.yelpcorp.com/i/wk9Hw3Wwl9L9T7fwRwX86Z8x0LrQmG4d.html
HPA works: https://fluffy.yelpcorp.com/i/RjJZ5DPbWSvS31Ls9l81DJ3MKhkg1V42.html

### CODE REVIEW

File: src/fullerite/collector/hpa_metrics.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		return 0, err
+	}
+	utilization, ok := result["utilization"].(string)
+	if !ok {
+		return 0, fmt.Errorf("\"utilization\" field not found or not a string")
+	}
+	return strconv.ParseFloat(utilization, 64)
+}
+
+// parseUWSGIMetrics return the percentage of none idle workers.
```
Comment (EvanKrall) [inline src/fullerite/collector/hpa_metrics.go]: non-idle

File: src/fullerite/collector/hpa_metrics.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		}
+		if status != "idle" {
+			activeWorker++
+		}
+	}
+	utilization = float64(activeWorker) / float64(totalWorker)
+	return utilization, err
+}
+
+// buildHPAMetric build a new Metrics.
```
Comment (EvanKrall) [inline src/fullerite/collector/hpa_metrics.go]: s/Metrics/Metric/, though this is pretty obvious from the return value.

File: src/fullerite/collector/hpa_metrics.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+	if additionalDimensions, exists := configMap["additionalDimensions"]; exists {
+		d.additionalDimensions = config.GetAsMap(additionalDimensions)
+	}
+
+	d.configureCommonParams(configMap)
+}
+
+// Collect Ping kubelet for pod specs. Iterates all pods, and collect http or uwsgi metrics
+// if all containers in the pod are healthy, and if there is "autoscaling"="http"/"uwsgi" in
```
Comment (EvanKrall) [inline src/fullerite/collector/hpa_metrics.go]: s/healthy/ready/

File: src/fullerite/collector/hpa_metrics.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+	for i := range podList.Items {
+		go d.CollectMetricsForPod(&podList.Items[i])
+	}
+}
+
+// CollectMetricsForPod collect http or uwsgi metrics if all containers in the pod are healthy,
+// and if there is "autoscaling"="http"/"uwsgi" in the annotation.
+func (d *HPAMetrics) CollectMetricsForPod(pod *corev1.Pod) {
+	metricsName, annotationPresent := pod.GetAnnotations()[autoscalingAnnotation]
+	if !annotationPresent || !d.allContainersAreReady(pod) {
```
[the remark this answers is not in the document]
Comment (EmanekaT) [reply src/fullerite/collector/hpa_metrics.go]: @EvanKrall I think hacheck does similar kind of checking. So we don't need to register the status here.

**Comment by EmanekaT** [comment]
@rahulravindran0108 This collector only emits metrics for instances that enabled http/uwsgi autoscaling on kubernetes. I think we only have 1 or 2 instances that actually need it, so the DPM would be negligible before we migrate more workloads to kubernetes. That will happen at least after Q1 2020.

**Pull request merged**